errs/error.go

```diff
@@ -0,0 +1,197 @@
+type StatusCoder interface {
+	StatusCode() int
+}
+
+// StackTracer must be by those errors that return an stack trace.
+type StackTracer interface {
+	StackTrace() errors.StackTrace
+}
+
+// Error represents the CA API errors.
+type Error struct {
+	Status int
+	Err    error
+	Msg    string
+}
+
+// ErrorResponse represents an error in JSON format.
+type ErrorResponse struct {
+	Status  int    `json:"status"`
+	Message string `json:"message"`
+}
+
+// Cause implements the errors.Causer interface and returns the original error.
+func (e *Error) Cause() error {
+	return e.Err
+}
+
+// Error implements the error interface and returns the error string.
+func (e *Error) Error() string {
+	return e.Err.Error()
+}
+
+// StatusCode implements the StatusCoder interface and returns the HTTP response
+// code.
+func (e *Error) StatusCode() int {
+	return e.Status
+}
+
+// Message returns a user friendly error, if one is set.
+func (e *Error) Message() string {
+	if len(e.Msg) > 0 {
+		return e.Msg
+	}
+	return e.Err.Error()
+}
+
+// WithMessage sets the user friendly message for a given error.
+func WithMessage(e error, m string) error {
+	if err, ok := e.(*Error); ok {
+		err.Msg = m
+		return err
+	}
+	return &Error{
+		Status: http.StatusInternalServerError,
+		Err:    e,
+		Msg:    m,
+	}
+}
+
+// Wrap returns an error annotating err with a stack trace at the point Wrap is
+// called, and the supplied message. If err is nil, Wrap returns nil.
+func Wrap(e error, m string) error {
+	if e == nil {
+		return nil
+	}
+	if err, ok := e.(*Error); ok {
+		err.Err = errors.Wrap(err.Err, m)
+		return err
+	}
+	return errors.Wrap(e, m)
+}
+
+// Wrapf returns an error annotating err with a stack trace at the point Wrap is
+// called, and the supplied message. If err is nil, Wrap returns nil.
+func Wrapf(e error, format string, args ...interface{}) error {
+	if e == nil {
+		return nil
+	}
+	if err, ok := e.(*Error); ok {
+		err.Err = errors.Wrapf(err.Err, format, args...)
+		return err
+	}
+	return errors.Wrapf(e, format, args...)
+}
+
+// MarshalJSON implements json.Marshaller interface for the Error struct.
+func (e *Error) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&ErrorResponse{Status: e.Status, Message: http.StatusText(e.Status)})
+}
+
+// Format implements the fmt.Formatter interface.
+func (e *Error) Format(f fmt.State, c rune) {
+	if err, ok := e.Err.(fmt.Formatter); ok {
+		err.Format(f, c)
+		return
+	}
+	fmt.Println(e.Err.Error())
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface for the Error struct.
+func (e *Error) UnmarshalJSON(data []byte) error {
+	var er ErrorResponse
+	if err := json.Unmarshal(data, &er); err != nil {
+		return err
+	}
+	e.Status = er.Status
+	e.Err = fmt.Errorf(er.Message)
+	return nil
+}
+
+// NewError returns a new Error. If the given error implements the StatusCoder
+// interface we will ignore the given status.
+func NewError(status int, err error) error {
```

Comment:
Perhaps we can add a variadic parameter that allows setting the Message, or create a new method that allows that, something like `NewErrorMsg(status, err, msg)`

Reply:
Adding here proposal discussed over slack:
```go
type Option func(e *Error) // or type Option func(e *Error) error

func New(status int, err error, opts ...Option) {}
func Wrap(status int, err error, msg string, opts ...Options) {}
// The args before the first option will be the arguments for the format string.
func Wrapf(status int, err error, format string, args ...interface{}) {}
```